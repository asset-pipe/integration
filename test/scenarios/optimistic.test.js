'use strict';

const { resolve } = require;
const { spawn } = require('child_process');
const Client = require('@asset-pipe/client');
const buildServerUri = 'http://127.0.0.1:7100';
const supertest = require('supertest');
const request = supertest(buildServerUri);
const { hashArray } = require('../../utils');
const {
    endWorkers,
} = require('../../node_modules/@asset-pipe/server/lib/utils');

let server;

function startServer(env = 'development') {
    return new Promise((resolve, reject) => {
        server = spawn('./node_modules/.bin/asset-pipe-server', [], {
            env: Object.assign({}, process.env, {
                NODE_ENV: env,
            }),
        });
        server.stdout.once('data', () => resolve());
        server.stderr.once('data', () => resolve());
        server.once('error', err => reject(err));
    });
}

function stopServer() {
    return new Promise(resolve => {
        server.once('close', () => resolve());
        server.kill();
    });
}

beforeEach(() => {
    jest.setTimeout(20000);
    return startServer();
});

afterEach(() => stopServer());

afterAll(() => endWorkers());

test('scenario: 2 podlets come online after layout', async () => {
    expect.assertions(2);
    const client = new Client({ buildServerUri });

    // layout publishes instructions for js and css bundling
    await client.publishInstructions('layout', 'js', ['podlet1', 'podlet2']);
    await client.publishInstructions('layout', 'css', ['podlet1', 'podlet2']);

    // podlet 1 publishes assets
    const [p1js, p1css] = await Promise.all([
        await client.publishAssets('podlet1', [resolve('../../assets/a.js')]),
        await client.publishAssets('podlet1', [resolve('../../assets/a.css')]),
    ]);

    // podlet 2 publishes assets
    const [p2js, p2css] = await Promise.all([
        await client.publishAssets('podlet2', [resolve('../../assets/b.js')]),
        await client.publishAssets('podlet2', [resolve('../../assets/b.css')]),
    ]);

    const hash1 = await hashArray([p1js.id, p2js.id]);
    const { text: js } = await request.get(`/bundle/${hash1}.js`).expect(200);

    const hash2 = await hashArray([p1css.id, p2css.id]);
    const { text: css } = await request.get(`/bundle/${hash2}.css`).expect(200);

    expect(js).toMatchSnapshot();
    expect(css).toMatchSnapshot();
});

test('scenario: 2 podlets come online first then layout publishes after', async () => {
    expect.assertions(2);
    const client = new Client({ buildServerUri });

    // podlet 1 publishes assets
    const [p1js, p1css] = await Promise.all([
        await client.publishAssets('podlet1', [resolve('../../assets/a.js')]),
        await client.publishAssets('podlet1', [resolve('../../assets/a.css')]),
    ]);

    // podlet 2 publishes assets
    const [p2js, p2css] = await Promise.all([
        await client.publishAssets('podlet2', [resolve('../../assets/b.js')]),
        await client.publishAssets('podlet2', [resolve('../../assets/b.css')]),
    ]);

    // layout publishes instructions for js and css bundling
    await client.publishInstructions('layout', 'js', ['podlet1', 'podlet2']);
    await client.publishInstructions('layout', 'css', ['podlet1', 'podlet2']);

    const hash1 = await hashArray([p1js.id, p2js.id]);
    const { text: js } = await request.get(`/bundle/${hash1}.js`).expect(200);

    const hash2 = await hashArray([p1css.id, p2css.id]);
    const { text: css } = await request.get(`/bundle/${hash2}.css`).expect(200);

    expect(js).toMatchSnapshot();
    expect(css).toMatchSnapshot();
});

test('scenario: changes to podlets trigger rebundling', async () => {
    expect.assertions(4);
    const client = new Client({ buildServerUri });

    // podlet 1 publishes assets
    const [p1js, p1css] = await Promise.all([
        await client.publishAssets('podlet1', [resolve('../../assets/c.js')]),
        await client.publishAssets('podlet1', [resolve('../../assets/c.css')]),
    ]);

    // podlet 2 publishes assets
    const [p2js, p2css] = await Promise.all([
        await client.publishAssets('podlet2', [resolve('../../assets/d.js')]),
        await client.publishAssets('podlet2', [resolve('../../assets/d.css')]),
    ]);

    // layout publishes instructions for js and css bundling
    await client.publishInstructions('layout', 'js', ['podlet1', 'podlet2']);
    await client.publishInstructions('layout', 'css', ['podlet1', 'podlet2']);

    // podlet 1 republishes assets
    const [p1rebundlejs, p1rebundlecss] = await Promise.all([
        await client.publishAssets('podlet1', [resolve('../../assets/e.js')]),
        await client.publishAssets('podlet1', [resolve('../../assets/e.css')]),
    ]);

    const hash1 = await hashArray([p1js.id, p2js.id]);
    const { text: jsBundle1 } = await request
        .get(`/bundle/${hash1}.js`)
        .expect(200);
    expect(jsBundle1).toMatchSnapshot();

    const hash2 = await hashArray([p1css.id, p2css.id]);
    const { text: cssBundle1 } = await request
        .get(`/bundle/${hash2}.css`)
        .expect(200);
    expect(cssBundle1).toMatchSnapshot();

    const hash3 = await hashArray([p1rebundlejs.id, p2js.id]);
    const { text: jsBundle2 } = await request
        .get(`/bundle/${hash3}.js`)
        .expect(200);
    expect(jsBundle2).toMatchSnapshot();

    const hash4 = await hashArray([p1rebundlecss.id, p2css.id]);
    const { text: cssBundle2 } = await request
        .get(`/bundle/${hash4}.css`)
        .expect(200);
    expect(cssBundle2).toMatchSnapshot();
});

test('scenario: adding a podlet to bundle instructions generates a new bundle', async () => {
    expect.assertions(4);
    const client = new Client({ buildServerUri });

    // podlet 1 publishes assets
    const [p1js, p1css] = await Promise.all([
        await client.publishAssets('podlet1', [resolve('../../assets/a.js')]),
        await client.publishAssets('podlet1', [resolve('../../assets/a.css')]),
    ]);

    // podlet 2 publishes assets
    const [p2js, p2css] = await Promise.all([
        await client.publishAssets('podlet2', [resolve('../../assets/b.js')]),
        await client.publishAssets('podlet2', [resolve('../../assets/b.css')]),
    ]);

    // layout publishes instructions for js and css bundling of podlet1
    await client.publishInstructions('layout', 'js', ['podlet1']);
    await client.publishInstructions('layout', 'css', ['podlet1']);

    // layout publishes instructions for js and css bundling of both podlet1 and 2
    await client.publishInstructions('layout', 'js', ['podlet1', 'podlet2']);
    await client.publishInstructions('layout', 'css', ['podlet1', 'podlet2']);

    const hash1 = await hashArray([p1js.id]);
    const { text: jsBundle1 } = await request
        .get(`/bundle/${hash1}.js`)
        .expect(200);
    expect(jsBundle1).toMatchSnapshot();

    const hash2 = await hashArray([p1css.id]);
    const { text: cssBundle1 } = await request
        .get(`/bundle/${hash2}.css`)
        .expect(200);
    expect(cssBundle1).toMatchSnapshot();

    const hash3 = await hashArray([p1js.id, p2js.id]);
    const { text: jsBundle2 } = await request
        .get(`/bundle/${hash3}.js`)
        .expect(200);
    expect(jsBundle2).toMatchSnapshot();

    const hash4 = await hashArray([p1css.id, p2css.id]);
    const { text: cssBundle2 } = await request
        .get(`/bundle/${hash4}.css`)
        .expect(200);
    expect(cssBundle2).toMatchSnapshot();
});

test('scenario: removing a podlet from bundle instructions generates a new bundle', async () => {
    expect.assertions(4);
    const client = new Client({ buildServerUri });

    // podlet 1 publishes assets
    const [p1js, p1css] = await Promise.all([
        await client.publishAssets('podlet1', [resolve('../../assets/a.js')]),
        await client.publishAssets('podlet1', [resolve('../../assets/a.css')]),
    ]);

    // podlet 2 publishes assets
    const [p2js, p2css] = await Promise.all([
        await client.publishAssets('podlet2', [resolve('../../assets/b.js')]),
        await client.publishAssets('podlet2', [resolve('../../assets/b.css')]),
    ]);

    // layout publishes instructions for js and css bundling of both podlet1 and 2
    await client.publishInstructions('layout', 'js', ['podlet1', 'podlet2']);
    await client.publishInstructions('layout', 'css', ['podlet1', 'podlet2']);

    // layout publishes instructions for js and css bundling of podlet1
    await client.publishInstructions('layout', 'js', ['podlet1']);
    await client.publishInstructions('layout', 'css', ['podlet1']);

    const hash3 = await hashArray([p1js.id, p2js.id]);
    const { text: jsBundle2 } = await request
        .get(`/bundle/${hash3}.js`)
        .expect(200);
    expect(jsBundle2).toMatchSnapshot();

    const hash4 = await hashArray([p1css.id, p2css.id]);
    const { text: cssBundle2 } = await request
        .get(`/bundle/${hash4}.css`)
        .expect(200);
    expect(cssBundle2).toMatchSnapshot();

    const hash1 = await hashArray([p1js.id]);
    const { text: jsBundle1 } = await request
        .get(`/bundle/${hash1}.js`)
        .expect(200);
    expect(jsBundle1).toMatchSnapshot();

    const hash2 = await hashArray([p1css.id]);
    const { text: cssBundle1 } = await request
        .get(`/bundle/${hash2}.css`)
        .expect(200);
    expect(cssBundle1).toMatchSnapshot();
});
