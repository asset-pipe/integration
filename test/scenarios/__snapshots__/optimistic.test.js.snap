// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenario: 2 podlets come online after layout 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}]},{},[1,3]);
"
`;

exports[`scenario: 2 podlets come online after layout 2`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}"
`;

exports[`scenario: 2 podlets come online first then layout publishes after 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}]},{},[1,3]);
"
`;

exports[`scenario: 2 podlets come online first then layout publishes after 2`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}"
`;

exports[`scenario: adding a podlet to bundle instructions generates a new bundle 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}]},{},[1]);
"
`;

exports[`scenario: adding a podlet to bundle instructions generates a new bundle 2`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}"
`;

exports[`scenario: adding a podlet to bundle instructions generates a new bundle 3`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}]},{},[1,3]);
"
`;

exports[`scenario: adding a podlet to bundle instructions generates a new bundle 4`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}"
`;

exports[`scenario: changes to podlets trigger rebundling 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(5, 6);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}]},{},[1,3]);
"
`;

exports[`scenario: changes to podlets trigger rebundling 2`] = `
"body {
    margin: 0;
    padding: 0;
}

.c-title {
    color: green;
}

body {
    margin: 0;
    padding: 0;
}"
`;

exports[`scenario: changes to podlets trigger rebundling 3`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],2:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(10, 6);

},{\\"./d\\":1}],3:[function(require,module,exports){
arguments[4][1][0].apply(exports,arguments)
},{\\"dup\\":1}]},{},[2,3]);
"
`;

exports[`scenario: changes to podlets trigger rebundling 4`] = `
".e-title {
    color: black;
}

body {
    margin: 0;
    padding: 0;
}"
`;

exports[`scenario: full production setup 1`] = `"!function t(n,e,r){function o(u,c){if(!e[u]){if(!n[u]){var s=\\"function\\"==typeof require&&require;if(!c&&s)return s(u,!0);if(i)return i(u,!0);var f=new Error(\\"Cannot find module '\\"+u+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=e[u]={exports:{}};n[u][0].call(l.exports,function(t){var e=n[u][1][t];return o(e||t)},l,l.exports,t,n,e,r)}return e[u].exports}for(var i=\\"function\\"==typeof require&&require,u=0;u<r.length;u++)o(r[u]);return o}({1:[function(t,n,e){\\"use strict\\";t(\\"./d\\")(1,2)},{\\"./d\\":2}],2:[function(t,n,e){\\"use strict\\";n.exports=function(t,n){return t+n}},{}],3:[function(t,n,e){t(\\"./a\\")},{\\"./a\\":1}],4:[function(t,n,e){\\"use strict\\";t(\\"./d\\")(3,4)},{\\"./d\\":5}],5:[function(t,n,e){arguments[4][2][0].apply(e,arguments)},{dup:2}],6:[function(t,n,e){var r,o,i=n.exports={};function u(){throw new Error(\\"setTimeout has not been defined\\")}function c(){throw new Error(\\"clearTimeout has not been defined\\")}function s(t){if(r===setTimeout)return setTimeout(t,0);if((r===u||!r)&&setTimeout)return r=setTimeout,setTimeout(t,0);try{return r(t,0)}catch(n){try{return r.call(null,t,0)}catch(n){return r.call(this,t,0)}}}!function(){try{r=\\"function\\"==typeof setTimeout?setTimeout:u}catch(t){r=u}try{o=\\"function\\"==typeof clearTimeout?clearTimeout:c}catch(t){o=c}}();var f,l=[],a=!1,p=-1;function h(){a&&f&&(a=!1,f.length?l=f.concat(l):p=-1,l.length&&d())}function d(){if(!a){var t=s(h);a=!0;for(var n=l.length;n;){for(f=l,l=[];++p<n;)f&&f[p].run();p=-1,n=l.length}f=null,a=!1,function(t){if(o===clearTimeout)return clearTimeout(t);if((o===c||!o)&&clearTimeout)return o=clearTimeout,clearTimeout(t);try{o(t)}catch(n){try{return o.call(null,t)}catch(n){return o.call(this,t)}}}(t)}}function m(t,n){this.fun=t,this.array=n}function y(){}i.nextTick=function(t){var n=new Array(arguments.length-1);if(arguments.length>1)for(var e=1;e<arguments.length;e++)n[e-1]=arguments[e];l.push(new m(t,n)),1!==l.length||a||s(d)},m.prototype.run=function(){this.fun.apply(null,this.array)},i.title=\\"browser\\",i.browser=!0,i.env={},i.argv=[],i.version=\\"\\",i.versions={},i.on=y,i.addListener=y,i.once=y,i.off=y,i.removeListener=y,i.removeAllListeners=y,i.emit=y,i.prependListener=y,i.prependOnceListener=y,i.listeners=function(t){return[]},i.binding=function(t){throw new Error(\\"process.binding is not supported\\")},i.cwd=function(){return\\"/\\"},i.chdir=function(t){throw new Error(\\"process.chdir is not supported\\")},i.umask=function(){return 0}},{}],7:[function(t,n,e){arguments[4][4][0].apply(e,arguments)},{\\"./d\\":8,dup:4}],8:[function(t,n,e){arguments[4][2][0].apply(e,arguments)},{dup:2}],9:[function(t,n,e){\\"use strict\\";t(\\"./b\\");spy(\\"g\\")},{\\"./b\\":7}],10:[function(t,n,e){(function(t){console.log(\\"In prod!!\\")}).call(this,t(\\"./../../asset-pipe-client/node_modules/process/browser.js\\"))},{\\"./../../asset-pipe-client/node_modules/process/browser.js\\":6}]},{},[3,4,9,10]);"`;

exports[`scenario: full production setup 2`] = `"!function n(t,e,r){function o(u,c){if(!e[u]){if(!t[u]){var s=\\"function\\"==typeof require&&require;if(!c&&s)return s(u,!0);if(i)return i(u,!0);var f=new Error(\\"Cannot find module '\\"+u+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var a=e[u]={exports:{}};t[u][0].call(a.exports,function(n){var e=t[u][1][n];return o(e||n)},a,a.exports,n,t,e,r)}return e[u].exports}for(var i=\\"function\\"==typeof require&&require,u=0;u<r.length;u++)o(r[u]);return o}({1:[function(n,t,e){\\"use strict\\";n(\\"./d\\")(1,2)},{\\"./d\\":2}],2:[function(n,t,e){\\"use strict\\";t.exports=function(n,t){return n+t}},{}],3:[function(n,t,e){n(\\"./a\\")},{\\"./a\\":1}],4:[function(n,t,e){arguments[4][1][0].apply(e,arguments)},{\\"./d\\":5,dup:1}],5:[function(n,t,e){arguments[4][2][0].apply(e,arguments)},{dup:2}],6:[function(n,t,e){arguments[4][3][0].apply(e,arguments)},{\\"./a\\":4,dup:3}],7:[function(n,t,e){\\"use strict\\";n(\\"./d\\")(3,4)},{\\"./d\\":8}],8:[function(n,t,e){arguments[4][2][0].apply(e,arguments)},{dup:2}],9:[function(n,t,e){var r,o,i=t.exports={};function u(){throw new Error(\\"setTimeout has not been defined\\")}function c(){throw new Error(\\"clearTimeout has not been defined\\")}function s(n){if(r===setTimeout)return setTimeout(n,0);if((r===u||!r)&&setTimeout)return r=setTimeout,setTimeout(n,0);try{return r(n,0)}catch(t){try{return r.call(null,n,0)}catch(t){return r.call(this,n,0)}}}!function(){try{r=\\"function\\"==typeof setTimeout?setTimeout:u}catch(n){r=u}try{o=\\"function\\"==typeof clearTimeout?clearTimeout:c}catch(n){o=c}}();var f,a=[],l=!1,p=-1;function d(){l&&f&&(l=!1,f.length?a=f.concat(a):p=-1,a.length&&h())}function h(){if(!l){var n=s(d);l=!0;for(var t=a.length;t;){for(f=a,a=[];++p<t;)f&&f[p].run();p=-1,t=a.length}f=null,l=!1,function(n){if(o===clearTimeout)return clearTimeout(n);if((o===c||!o)&&clearTimeout)return o=clearTimeout,clearTimeout(n);try{o(n)}catch(t){try{return o.call(null,n)}catch(t){return o.call(this,n)}}}(n)}}function m(n,t){this.fun=n,this.array=t}function y(){}i.nextTick=function(n){var t=new Array(arguments.length-1);if(arguments.length>1)for(var e=1;e<arguments.length;e++)t[e-1]=arguments[e];a.push(new m(n,t)),1!==a.length||l||s(h)},m.prototype.run=function(){this.fun.apply(null,this.array)},i.title=\\"browser\\",i.browser=!0,i.env={},i.argv=[],i.version=\\"\\",i.versions={},i.on=y,i.addListener=y,i.once=y,i.off=y,i.removeListener=y,i.removeAllListeners=y,i.emit=y,i.prependListener=y,i.prependOnceListener=y,i.listeners=function(n){return[]},i.binding=function(n){throw new Error(\\"process.binding is not supported\\")},i.cwd=function(){return\\"/\\"},i.chdir=function(n){throw new Error(\\"process.chdir is not supported\\")},i.umask=function(){return 0}},{}],10:[function(n,t,e){arguments[4][7][0].apply(e,arguments)},{\\"./d\\":11,dup:7}],11:[function(n,t,e){arguments[4][2][0].apply(e,arguments)},{dup:2}],12:[function(n,t,e){\\"use strict\\";n(\\"./b\\");spy(\\"g\\")},{\\"./b\\":10}],13:[function(n,t,e){(function(n){console.log(\\"In prod!!\\")}).call(this,n(\\"./../../asset-pipe-client/node_modules/process/browser.js\\"))},{\\"./../../asset-pipe-client/node_modules/process/browser.js\\":9}]},{},[3,7,12,13,6]);"`;

exports[`scenario: full production setup 3`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}

.error {
    color: red;
}

.e-title {
    color: black;
}"
`;

exports[`scenario: full production setup 4`] = `
"body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}

.error {
    color: red;
}

.e-title {
    color: black;
}

body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}"
`;

exports[`scenario: removing a podlet from bundle instructions generates a new bundle 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}]},{},[1,3]);
"
`;

exports[`scenario: removing a podlet from bundle instructions generates a new bundle 2`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}"
`;

exports[`scenario: removing a podlet from bundle instructions generates a new bundle 3`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}]},{},[1]);
"
`;

exports[`scenario: removing a podlet from bundle instructions generates a new bundle 4`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}"
`;
